### DISCUSSION TITLE
How to convert 10bit HEIF to other 10bit lossless format?

### DISCUSSION THREAD

**Discussion by sagz** [ImageMagick/ImageMagick#3883] [Help]
Hi, the new Oppo Find X3 Pro creates 10bit HEIC files. These photos aren't supported by many other apps like Lightroom Classic, and Google Photos. I am wondering if it'll be possible to convert the file to a RAW format like DNG with an 8bit JPG embedded in it. This file would be supported by Google Photos when using the Original Quality tier too.

Here's a sample file. I had to zip it as Github doesn't allow `*.heic` files 🤦🏽: [IMG20210627090415.heic.zip](https://github.com/ImageMagick/ImageMagick/files/6728179/IMG20210627090415.heic.zip)

I tried `convert IMG20210627090415.heic IMG20210627090415.dng` but that didn't preserve the color profile or the 10bit depth.

Here's the (rather long) mediainfo dump of the file:
```
General
Complete name                            : IMG20210627090415.heic
Format                                   : heic
Codec ID                                 : heic (mif1/heic)
File size                                : 3.32 MiB
IsTruncated                              : Yes

Image #1
ID                                       : 10000
Format                                   : HEVC
Format/Info                              : High Efficiency Video Coding
Format profile                           : Main 10@L5@Main
Codec ID                                 : hvc1
Codec ID/Info                            : High Efficiency Video Coding
Width                                    : 512 pixels
Height                                   : 512 pixels
Color space                              : YUV
Chroma subsampling                       : 4:2:0
Bit depth                                : 10 bits
Stream size                              : 57.4 KiB (2%)
Default                                  : No
Color primaries                          : BT.601 PAL
Transfer characteristics                 : sRGB/sYCC
Matrix coefficients                      : BT.470 System B/G
Standard                                 : NTSC
Color range                              : Full
Derived                                  : 10048
Codec configuration box                  : hvcC

Image #2
...
...
Image #48
ID                                       : 10047
Format                                   : HEVC
Format/Info                              : High Efficiency Video Coding
Format profile                           : Main 10@L5@Main
Codec ID                                 : hvc1
Codec ID/Info                            : High Efficiency Video Coding
Width                                    : 512 pixels
Height                                   : 512 pixels
Color space                              : YUV
Chroma subsampling                       : 4:2:0
Bit depth                                : 10 bits
Stream size                              : 119 KiB (3%)
Default                                  : No
Color primaries                          : BT.601 PAL
Transfer characteristics                 : sRGB/sYCC
Matrix coefficients                      : BT.470 System B/G
Standard                                 : NTSC
Color range                              : Full
Derived                                  : 10048
Codec configuration box                  : hvcC

Image #49
ID                                       : 10048
Format                                   : Grid
Width                                    : 4 096 pixels
Height                                   : 3 072 pixels
Stream size                              : 8.00 Bytes (0%)
Derived for                              : 10000,10001,10002,10003,10004,10005,10006,10007,10008,10009,10010,10011,10012,10013,10014,10015,10016,10017,10018,10019,10020,10021,10022,10023,10024,10025,10026,10027,10028,10029,10030,10031,10032,10033,10034,10035,10036,10037,10038,10039,10040,10041,10042,10043,10044,10045,10046,10047
Describes                                : 10049
```

**Comment by snibgo** [comment]
You seem to be asking at least three questions.

(1) Can IM read your 10-bit HEIC files? I don't know. Try it and see.

(2) Can IM write DNG files? Not as far as I know. However, DNG is a form of TIFF, and IM can write TIFF files, and conversion from that to a DNG is fairly simple. I don't know why you would want this, unless you want to directly manipulate raw camera channels.

(3) Can IM write TIFF files with 10 bits per channel? Yes: "magick in.png -depth 10 out.tiff".

**Reply by sagz** [reply]
Thanks for the quick response, `snibgo`.

For #1:
```
-> identify IMG20210627090415.heic                       
IMG20210627090415.heic HEIC 4096x3072 4096x3072+0+0 10-bit sRGB 0.000u 0:00.005
```
Does this mean IM can 'read' 10-bit HEIC?

For #2, it does seem to support writing DNG files as the command `magick convert IMG20210627090415.heic IMG20210627090415.dng` does work, but weirdly leads to smaller files than the HEIC and only in 8bit. But I'm not married to DNG here and can do PNG as that also supports higher-than-8-bit.

Which brings me to #3, where `magick convert IMG20210627090415.heic -depth 10 IMG20210627090415.png` seems to have worked but I need to test a bit more to ensure I didn't drop something in the conversion. Why doesn't IM default to the source file's depth? Without your help I wouldn't have known to add `-depth 10`. Which makes me wonder what are some other flags I should add to ensure they don't get 'dropped'....

**Reply by fmw42** [reply]
I do not know if Imagemagick can read 10-bit HEIC, yet. But note that with IM 7, use magick, not magick convert when processing images. When saving to DNG, you may just be getting TIFF as the result.

**Reply by sagz** [reply]
Thanks `fmw42`, how do I verify if IM can read 10-bt HEIC?

`magick identify -verbose file.heic` does seem to work and output stuff seemingly as if it can read it...

**Comment by snibgo** [comment]
For #1, I suggest you try converting to a common format such as PNG:
```
magick IMG20210627090415.heic x.png
```
... then try viewing x.png in any viewer. Does it look correct? It doesn't for me, but my IM is old.

For #2: Confusingly, if you try to write to a format that IM doesn't recognise, IM won't complain. But the output won't be what you want. Instead, try:
```
magick -list format
```
This will include a line like this:
```
DNG  DNG       r--   Digital Negative
```
So my version of IM can read but not write DNG files.

10 bits/channel/pixel is an unusual format for writing files. If you plan on doing any further processing, I suggest using at least 16 bits.

**Reply by sagz** [reply]
` ➜ magick convert IMG20210627090415.heic IMG20210627090415.png`
reduces it to 8bit :(
so does ` ➜ magick convert IMG20210627090415.heic png00:IMG20210627090415.png`


One thing I don't understand when doing `magick identify -verbose img.heic` is this excerpt:
```
...
Depth: 10/8-bit
  Channel depth:
    Red: 8-bit
    Green: 8-bit
    Blue: 8-bit
  Channel statistics:
    Pixels: 12582912
    Red:
      min: 0  (0)
      max: 1023 (1)
...
```

What is `10/8-bit` mean? How is `Channel depth` of each color at `8-bit` but the `min`-`max` range is 1024 values (requiring 10 bits)?

**Comment by snibgo** [comment]
As fmw42 said above, I suggest using "magick" alone to get v7 syntax, not "magick convert" which gives v6 syntax.

I suspect that means that the file recorded 10-bit pixel values, but the bottom two bits were always the same as the top two bits.

To test this theory, what is the output of the following?
```
magick IMG20210627090415.heic -channel RGB -separate -append -unique-colors info:
```
The result width is the number of unique values across all channels. I suspect this is 256 or less for your image.

**Reply by fmw42** [reply]
Saving as RGB (as opposed to sRGB) most likely makes the image darker and compresses colors into the darker range. That may account for the 10/8-bit result. Why are you converting to RGB and not sRGB?

**Reply by sagz** [reply]
```
➜ magick IMG20210627090415.heic -channel RGB -separate -append -unique-colors info:
IMG20210627090415.heic HEIC 256x1 256x0+0+0 10-bit Gray 0.000u 0:00.005
```

**Comment by snibgo** [comment]
As I suspected: your values are from 0 to 1023, but there are only 256 different values, and only the top 8 bits are significant.

@fmw42: sagz isn't converting to RGB instead of sRGB. I think you are confusing two threads.

**Reply by sagz** [reply]
Ah. So the camera is claiming to be a 10bit image? sigh.

```
➜ dcraw -6 IMG20210627144016.dng
➜ magick IMG20210627144016.ppm -channel RGB -separate -append -unique-colors info:
IMG20210627144016.ppm PPM 38700x1 38700x0+0+0 16-bit Gray 72MiB 0.000u 0:00.002
```
Trying it on a RAW-mode photo from the camera gets me a higher value of `38700`.... probably too high?

At this point, your very first breakdown of questions about `#1 does IM even support 10b HEIF?` is the likely culprit for all this confusion. Is there a way to find out for sure? (apart from all the `identify` and `info:` commands we've run throughout this thead.


Thanks a lot for your help throughout this, `snibgo`!

**Comment by snibgo** [comment]
To record 38700 different values, we need 16 bits, of course.

The test showing 256 values from IMG20210627090415.heic is a measure after IM has read the values, and the HEIC library or IM may have stripped bits from the values. And the camera might record 256 possible value in each channel, but the 3x3 matrix multiplication to convert camera primaries to sRGB (or other) primaries can increase that number. For example, my camera records 14 bits/channel/pixel, but the camera matrix generates more than 32768 values per channel.

If you have a raw file that dcraw can read, and this file hasn't been manipulated by the camera, then:
```
dcraw -v -W -o 0 -6 -r 1 1 1 1 -g 1 0 -D -d -T -O x.tiff yourfile.heic
```
... where dcraw is as distributed with ImageMagick. Then find how many unique colours there are, as above, or you can remove the "-append" to count the colours in each channel:
```
magick x.tiff -channel RGB -separate -unique-colors info:
```

According to https://www.dxomark.com/oppo-find-x3-pro-camera-preview/ :

>  ... images on the Find X3 Pro can be saved in 10-bit HEIF format to include approximately one billion colors.

But then it says:
> Also similar to recent iPhones, the Oppo device uses the Display P3 color space when saving 10-bit images.

... which implies that if you want 10-bit images, they have to be processed to Display P3, and you can't have raw 10-bit images.

You ask:
> Is there a way to find out [whether IM supports 10-bt HEIC] for sure?

A developer may answer. A scan through IM or libheic source code might give a clue.

**Reply by dlemstra** [reply]
There is no real 10 bit support. The library that we use cannot properly read 10 bit images at this point in time. The images are scaled to 8 bit when they are being read.

**Reply by sagz** [reply]
Any way/process I can request support as a feature request?

Thank you for clarifying that IM doesn't actually support 10b HEIC, `dlemstra`!

**Reply by dlemstra** [reply]
I don't know if this is a problem for us or a libheif problem (see #2415). We will need to investigate this but I don't think that will happen anytime soon.
